subsys/bluetooth/audio/cap_initiator.c

```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+				atomic_set_bit(active_procedure.proc_state_flags,
+					       CAP_UNICAST_PROCEDURE_STATE_ABORTED);
+			} else {
+				(void)memset(&active_procedure, 0,
+					     sizeof(active_procedure));
+			}
+
+			return err;
+		}
+
+		active_procedure.stream_initiated++;
+	}
+
+	return 0;
+}
+
+static void cap_initiator_unicast_audio_start_complete(void)
+{
+	struct bt_audio_unicast_group *unicast_group;
+	struct bt_conn *failed_conn;
+	int err;
+
+	/* All streams in the procedure share the same unicast group, so we just
+	 * use the reference from the first stream
+	 */
+	unicast_group = active_procedure.streams[0]->bap_stream.unicast_group;
+	failed_conn = active_procedure.failed_conn;
+	err = active_procedure.err;
+
+	(void)memset(&active_procedure, 0, sizeof(active_procedure));
+	if (cap_cb != NULL && cap_cb->unicast_start_complete != NULL) {
+		cap_cb->unicast_start_complete(unicast_group, err, failed_conn);
+	}
+}
+
 int bt_cap_initiator_unicast_audio_start(const struct bt_cap_unicast_audio_start_param *param,
-					 struct bt_audio_unicast_group **unicast_group)
+					 struct bt_audio_unicast_group *unicast_group)
 {
-	return -ENOSYS;
+	if (atomic_test_bit(active_procedure.proc_state_flags,
+			    CAP_UNICAST_PROCEDURE_STATE_ACTIVE)) {
+		LOG_DBG("A CAP procedure is already in progress");
+
+		return -EBUSY;
+	}
+
+	CHECKIF(unicast_group == NULL) {
+		LOG_DBG("unicast_group is NULL");
+		return -EINVAL;
+	}
+
+	if (!valid_unicast_audio_start_param(param, unicast_group)) {
+		return -EINVAL;
+	}
+
+	active_procedure.unicast_group = unicast_group;
+
+	return cap_initiator_unicast_audio_configure(param);
 }
 
-int bt_cap_initiator_unicast_audio_update(struct bt_audio_unicast_group *unicast_group,
-					  uint8_t meta_count,
-					  const struct bt_codec_data *meta)
+void bt_cap_initiator_codec_configured(struct bt_cap_stream *cap_stream)
 {
-	return -ENOSYS;
+	struct bt_conn *conns[MIN(CONFIG_BT_MAX_CONN,
+				  CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT)];
+	struct bt_audio_unicast_group *unicast_group;
+
+	if (!cap_stream_in_active_procedure(cap_stream)) {
+		/* State change happened outside of a procedure; ignore */
+		return;
+	}
+
+	if (active_procedure.subproc_type == CAP_UNICAST_SUBPROCEDURE_TYPE_RELEASE) {
+		/* When releasing a stream, it may go into the codec configured state if
+		 * the unicast server caches the configuration - We treat it as a release
+		 */
+		bt_cap_initiator_released(cap_stream);
+		return;
+	} else if (active_procedure.subproc_type != CAP_UNICAST_SUBPROCEDURE_TYPE_CODEC_CONFIG) {
+		/* Unexpected callback - Abort */
+		active_procedure.err = -EBADMSG;
+		atomic_set_bit(active_procedure.proc_state_flags,
+			       CAP_UNICAST_PROCEDURE_STATE_ABORTED);
+	} else {
+		active_procedure.stream_done++;
+
+		LOG_DBG("Stream %p configured (%zu/%zu streams done)",
+			cap_stream, active_procedure.stream_done,
+			active_procedure.stream_cnt);
+
+		if (active_procedure.stream_done < active_procedure.stream_cnt) {
```

Comment:
```suggestion
static bool cap_procedure_is_done(void)
{
return active_procedure.stream_done == active_procedure.stream_cnt;
}
```
use some kind of helper function to be reused in other places places